isarn-sketches-java/src/main/java/org/isarnproject/sketches/java/TDigest.java

```diff
@@ -0,0 +1,761 @@
+        if (j == 0) return j;
+        // x is to right of right-most cluster:
+        if (j == nclusters) return j - 1;
+        // x is between two clusters, return index of closest:
+        double dL = x - cent[j - 1];
+        double dR = cent[j] - x;
+        return (dL < dR) ? (j - 1) : j;
+    }
+
+    /** Obtain the number of clusters in this t-digest 
+     * @return the number of clusters in this t-digest
+     */
+    public final int size() {
+        return nclusters;
+    }
+
+    /** Obtain the total mass sampled by this t-digest
+     * @return the total mass
+     */
+    public final double mass() {
+        return M;
+    }
+
+    /** Obtain the compression setting for this t-digest
+     * @return the compression setting
+     */
+    public final double getCompression() {
+        return C;
+    }
+
+    /** Obtain the maximum discrete setting for this t-digest
+     * @return the maximum discrete setting
+     */
+    public final int getMaxDiscrete() {
+        return maxDiscrete;
+    }
+
+    /** Obtain a reference to this t-digest's cluster center array.
+     * NOTE: this array is not safe to modify, and should be used only in "read-only" mode!
+     * @return a reference to the cluster center array
+     */
+    public final double[] getCent() {
+        return cent;
+    }
+
+    /** Obtain a reference to this t-digest's cluster mass array.
+     * NOTE: this array is not safe to modify, and should be used only in "read-only" mode!
+     * @return a reference to the cluster mass array
+     */
+    public final double[] getMass() {
+        return mass;
+    }
+
+    /** Obtain a reference to this t-digest's cumulative mass array.
+     * This array stores the cumulative masses of clusters in Fenwick Tree format.
+     * NOTE: this array is not safe to modify, and should be used only in "read-only" mode!
+     * @return a reference to the cumulative mass array
+     */
+    public final double[] getFT() {
+        return ftre;
+    }
+
+    /** Returns true if this t-digest is empty, false otherwise. */
+    public final boolean isEmpty() {
+        return nclusters == 0;
+    }
+
+    @Override
+    public String toString() {
+        StringBuilder sb = new StringBuilder("TDigest(");
+        for (int j = 0; j < nclusters; ++j) {
+            if (j > 25) {
+                sb.append(" ...");
+                break;
+            }
+            if (j > 0) sb.append(", ");
+            sb.append(cent[j])
+                .append(" -> (")
+                .append(mass[j])
+                .append(", ")
+                .append(ftSum(j))
+                .append(")");
+        }
+        sb.append(")");
+        return sb.toString();
+    }
+
+    /**
+     * Perform a random sampling from the distribution as sketched by this t-digest, in
+     * "probability density" mode.
+     * @return A random number sampled from the sketched distribution
+     */
+    public final double samplePDF() {
```

Comment:
I'd like to see a way to pass in a uniform PRNG here -- maybe as an optional argument, falling back to `ThreadLocalRandom` otherwise

Reply:
Makes sense